### PR TITLE
Issue: Cached query captures object in closure

### PR DESCRIPTION
**Pull request by SergeiPavlov**
It is expected that QueryCache should not keep references to any objects from closure.
Because cache record lifetime may be indefinitely long

### CODE REVIEW

File: Orm/Xtensive.Orm/Reflection/WellKnownTypes.cs [hunk 1]
```diff
@@ -73,6 +74,7 @@ internal static class WellKnownTypes
 
     public static readonly Type ByteArray = typeof(byte[]);
     public static readonly Type ObjectArray = typeof(object[]);
+    public static readonly Type IReadOnlyList = typeof(IReadOnlyList<>);
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Reflection/WellKnownTypes.cs:77]: IReadOnlyListOfT ?

File: Orm/Xtensive.Orm/Orm/Internals/CompiledQueryRunner.cs [hunk 2]
```diff
@@ -192,6 +199,25 @@ private void AllocateParameterAndReplacer()
         }
         return null;
       });
+
+      return !closureType.Name.Contains("<>c__DisplayClass")            // 'DisplayClass' is generated class for captured objects
+        || closureType.GetFields().All(FieldIsSimple);
+    }
+
+    private static bool TypeIsSimple(Type type)
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Internals/CompiledQueryRunner.cs:207]: IsSimpleType ?

File: Orm/Xtensive.Orm/Orm/Internals/CompiledQueryRunner.cs [hunk 3]
```diff
@@ -192,6 +199,25 @@ private void AllocateParameterAndReplacer()
         }
         return null;
       });
+
+      return !closureType.Name.Contains("<>c__DisplayClass")            // 'DisplayClass' is generated class for captured objects
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Internals/CompiledQueryRunner.cs:203]: Take a look at ```TypeHelper.IsClosure()``` extension method. Is it what you wanted to achieve here?

Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Internals/CompiledQueryRunner.cs]: Yes

File: Orm/Xtensive.Orm/Orm/Internals/CompiledQueryRunner.cs (hunk 2 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Internals/CompiledQueryRunner.cs]: renamed

File: Orm/Xtensive.Orm/Reflection/WellKnownTypes.cs (hunk 1 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Reflection/WellKnownTypes.cs]: renamed

**Review by alex-kulakov — APPROVED** [review]

**Pull request merged**